src/server/main_service.cc

```diff
@@ -732,6 +737,51 @@ void Service::Multi(CmdArgList args, ConnectionContext* cntx) {
   return (*cntx)->SendOk();
 }
 
+void Service::Watch(CmdArgList args, ConnectionContext* cntx) {
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  // Skip if EXEC will already fail due previous WATCH.
+  if (exec_info.watched_dirty.load(memory_order_relaxed)) {
+    return (*cntx)->SendOk();
+  }
+
+  atomic_uint32_t keys_existed = 0;
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    ArgSlice largs = t->ShardArgsInShard(shard->shard_id());
+    for (auto k: largs) shard->db_slice().RegisterWatchedKey(k, &exec_info);
+
+    auto res = GenericFamily::OpExists(t->GetOpArgs(shard), largs);
+    keys_existed.fetch_add(res.value_or(0), memory_order_relaxed);
+    return OpStatus::OK;
+  };
+  cntx->transaction->ScheduleSingleHop(std::move(cb));
+
+  exec_info.watched_existed += keys_existed.load(memory_order_relaxed);
```

Comment:
why "plus" do we support "watch ... ; watch ... " sequences?  
because then our counting is probably wrong.

Reply:
>  do we support "watch ... ; watch ... " sequences?

Redis does

> because then our counting is probably wrong.

Why? I specifically changed the set to a vector so that repeated keys work

Reply:
ah, i see. in db_slice? 
then lets add a comment here that duplicated keys are counted in the same way as in db_slice.